dqc_us_rules/dqc_us_0041.py

```diff
@@ -0,0 +1,294 @@
+    prior_validate_disclosure_system = (
+        val.modelXbrl.modelManager.validateDisclosureSystem
+    )
+    val.modelXbrl.modelManager.validateDisclosureSystem = False
+    calculations_instance = (ModelXbrl.load(
+        val.modelXbrl.modelManager,
+        openFileSource(ugt_entry_xsd, cntlr),
+        _("built us-gaap calculations cache"))  # noqa
+    )
+    val.modelXbrl.modelManager.validateDisclosureSystem = (
+        prior_validate_disclosure_system
+    )
+
+    if calculations_instance is None:
+        val.modelXbrl.error(
+            "arelle:notLoaded",
+            _("US-GAAP calculations not loaded: %(file)s"),  # noqa
+           modelXbrl=val,
+            file=os.path.basename(ugt_entry_xsd)
+        )
+
+    else:
+        for defaultDimRel in calculations_instance.relationshipSet(
+                XbrlConst.dimensionDefault).modelRelationships:
+            if isinstance(
+                    defaultDimRel.fromModelObject,
+                    ModelConcept
+            ) and isinstance(
+                defaultDimRel.toModelObject,
+                ModelConcept
+            ):
+                from_name = defaultDimRel.fromModelObject.name
+                to_name = defaultDimRel.toModelObject.name
+                val.usgaapDefaultDimensions[from_name] = to_name
+        json_str = str(json.dumps(
+            val.usgaapDefaultDimensions,
+            ensure_ascii=False,
+            indent=0)
+        )  # might not be unicode in 2.7
+        # 2.7 gets unicode this way
+        saveFile(cntlr, ugt_default_dimensions_json_file, json_str)
+        calculations_instance.close()
+        del calculations_instance  # dereference closed modelXbrl
+    val.modelXbrl.profileStat(
+        _("build default dimensions cache"),  # noqa
+        time.time() - started_at
+    )
+
+
+def _setup_cache(val):
+    """
+    Builds the cache needed for dqc_us_0041. Should only have to build it the
+    first time
+
+    :param val: ValidateXbrl to check if it contains errors
+    :type val: :class:'~arelle.ValidateXbrl.ValidateXbrl'
+    :return: No explicit return, but loads the default dimensions of taxonomies
+        into memory
+    :rtype: None
+    """
+    val.linroleDefinitionIsDisclosure = (
+        re.compile(r"-\s+Disclosure\s+-\s", re.IGNORECASE)
+    )
+
+    val.linkroleDefinitionStatementSheet = (
+        re.compile(r"[^-]+-\s+Statement\s+-\s+.*", re.IGNORECASE)
+    )  # no restriction to type of statement
+
+    val.ugtNamespace = None
+    cntlr = val.modelXbrl.modelManager.cntlr
+
+    year = 2014
+    for ugt in ugtDocs:
+        ugt_default_dimensions_json_file = os.path.join(
+            os.path.dirname(__file__),
+            'resources',
+            'DQC_US_0041',
+            "_".join([str(year), "ugt-default-dimensions.json"])
+        )
+
+        _make_cache(
+            val,
+            ugt,
+            cntlr,
+            ugt_default_dimensions_json_file
+        )
+
+        year += 1
```

Comment:
I'm not sure I understand what this line is doing... I'm leaning toward nothing...